subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1720,25 +1755,35 @@ static int ztls_poll_update_ctx(struct net_context *ctx,
 			if (recv(pfd->fd, NULL, 0, ZSOCK_MSG_DONTWAIT) < 0 &&
 			    errno != EAGAIN) {
 				pfd->revents |= ZSOCK_POLLERR;
-				return 0;
+				goto next;
 			}
 
 			if (mbedtls_ssl_get_bytes_avail(&ctx->tls->ssl) > 0 ||
 			    sock_is_eof(ctx)) {
 				pfd->revents |= ZSOCK_POLLIN;
-				return 0;
+				goto next;
 			}
 
 			/* Received encrypted data, but still not enough
 			 * to decrypt it and return data through socket,
 			 * ask for retry.
 			 */
-			errno = EAGAIN;
-			return -1;
+
+			(*pev)->state = K_POLL_STATE_NOT_READY;
+			goto again;
 		}
 	}
 
 	return 0;
+
+next:
+	(*pev)++;
+	return 0;
+
+again:
+	(*pev)++;
+	errno = EAGAIN;
```

Comment:
As was mentioned previously, it's quite confusing to see errno assignment here. errno should be reserved for communication between system and userspace application. Part of system would rather communicate with -EAGAIN.